src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,3315 @@
+    StacksBlockId,
+};
+use crate::util::boot::boot_code_id;
+
+/// Affirmation map entries.  By building on a PoX-mined block,
+/// a PoB-mined block (in a PoX reward cycle),
+/// or no block in reward cycle _i_, a sortition's miner
+/// affirms something about the status of the ancestral anchor blocks.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub enum AffirmationMapEntry {
+    PoxAnchorBlockPresent,
+    PoxAnchorBlockAbsent,
+    Nothing,
+}
+
+impl AffirmationMapEntry {
+    pub fn parse(s: &str) -> Option<AffirmationMapEntry> {
+        if s.len() != 1 {
+            return None;
+        }
+
+        for chr in s.chars() {
+            let next = match chr {
+                'p' => AffirmationMapEntry::PoxAnchorBlockPresent,
+                'a' => AffirmationMapEntry::PoxAnchorBlockAbsent,
+                'n' => AffirmationMapEntry::Nothing,
+                _ => {
+                    return None;
+                }
+            };
+            return Some(next);
+        }
+        return None;
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct AffirmationMap {
+    affirmations: Vec<AffirmationMapEntry>,
+}
+
+impl fmt::Display for AffirmationMapEntry {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            AffirmationMapEntry::PoxAnchorBlockPresent => write!(f, "p"),
+            AffirmationMapEntry::PoxAnchorBlockAbsent => write!(f, "a"),
+            AffirmationMapEntry::Nothing => write!(f, "n"),
+        }
+    }
+}
+
+impl fmt::Debug for AffirmationMapEntry {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&format!("{}", &self))
+    }
+}
+
+impl fmt::Display for AffirmationMap {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "AM(")?;
+        for entry in self.affirmations.iter() {
+            write!(f, "{}", &entry)?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl fmt::Debug for AffirmationMap {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&format!("{}", &self))
+    }
+}
+
+impl AffirmationMap {
+    pub fn new(entries: Vec<AffirmationMapEntry>) -> AffirmationMap {
+        AffirmationMap {
+            affirmations: entries,
+        }
+    }
+
+    pub fn empty() -> AffirmationMap {
+        AffirmationMap {
+            affirmations: vec![],
+        }
+    }
+
+    pub fn at(&self, reward_cycle: u64) -> Option<AffirmationMapEntry> {
+        if reward_cycle >= self.affirmations.len() as u64 {
+            None
+        } else {
+            Some(self.affirmations[reward_cycle as usize])
+        }
+    }
+
+    pub fn push(&mut self, entry: AffirmationMapEntry) {
+        self.affirmations.push(entry)
+    }
+
+    pub fn pop(&mut self) -> Option<AffirmationMapEntry> {
+        self.affirmations.pop()
+    }
+
+    pub fn len(&self) -> usize {
+        self.affirmations.len()
+    }
+
+    pub fn reward_cycle(&self) -> u64 {
```

Comment:
A rustdoc would be helpful here -- what is `affirmation_map.reward_cycle()` trying to return?

Reply:
It's not used anywhere (anymore), so I'll just remove it instead.